### ISSUE TITLE
Incompatible with lein-parent

### ISSUE THREAD

**Issue by nicorikken** [weavejester/codox#160]
**Issue**
`lein-codox` is currently incompatible with `lein-parent`, as a result of using `leinjacker.deps/add-if-missing`.

**Analysis**
`lein-parent` enables inheritance of dependencies from a parent project (with `:dependencies` options) and version numbers of dependencies (with `:managed-dependencies` option). When only inheriting version numbers, the dependency vector is filled with version numbers `nil`.
Excerpt of `lein-pprint` output:
```clojure
...
:dependencies
 ([org.clojure/core.async nil]
  [io.pedestal/pedestal.service "0.5.1"]
  [io.pedestal/pedestal.route nil]
  [io.pedestal/pedestal.jetty nil]
  [com.taoensso/timbre "4.8.0"]
...
```
`leinjacker.deps/add-if-missing` starts by checking the correctness of the dependency definition using [`deb-spec?`](https://github.com/sattvik/leinjacker/blob/9360fe6b1ec66e6011d0d5f67d040d4870ecbaae/src/leinjacker/deps.clj#L6) This validates whether or not the second element is a `string`, which it is not. And so `codox` cannot be included as a dependency, and so `lein-codox` no longer works.

**Workaround**
1. Explicitly add the dependency:
```clojure
[codox "0.10.3"]
```
2. Add an alias to mimic the behavior of `lein-codox`:
```clojure
:aliases {"codox" ["run" "-m" "codox.main/generate-docs"]}
```

**Ways of fixing**
- Fixing or forking `leinjacker.deps` to also accept a version number of `nil`, perhaps if an option is provided.
- Surround the `add-if-missing` with a try-catch, and refrain from adding the dependency if the insert fails. It is then up to the user to insert the dependency themselves (as is done in the workaround).

**Comment by aroemers** [comment]
👍

**Comment by MatthewDarling** [comment]
When using `:managed-dependencies`, it's also valid to drop the second argument completely:

```clojure
(defproject foo
  :managed-dependencies [[org.clojure/clojure "1.8.0"]]
  :dependencies [[org.clojure/clojure]])
```

In this case, you still get an error in `add-if-missing`, but it fails the `(>= (count dep) 2)` check instead.

**Comment by MatthewDarling** [comment]
I've made a [fork of `leinjacker`](https://github.com/sattvik/leinjacker/compare/master...MatthewDarling:master) that supports `:managed-dependencies`. I've added tests for the new syntax, as well, based on my reading of the [managed deps doc](https://github.com/technomancy/leiningen/blob/master/doc/MANAGED_DEPS.md) and the [managed deps test project](https://github.com/technomancy/leiningen/blob/master/test_projects/managed-deps-snapshot/project.clj). Running things at the REPL, it seems to process my `project.clj` fine.

Testing it thoroughly has proven to be a problem. The tests in `leinjacker.deps-test` pass when I run them from a REPL. However, the new tests fail when run through `lein test leinjacker.deps-test`, almost as if it's using the older behaviour somehow... And `lein test` for all namespaces produces a lot of failures, plus asks you to install Leiningen 1, and then `lein upgrade 1.7.1` fails because it tries to download Clojure from an outdated location.

Since running the existing tests is problematic, I figured I could at least test the portions that `codox` uses and go from there. I had trouble with that too, though, because I'm new to Lein plugin development. Tried a few things, like running a REPL for `lein-codox` and locally installing versions, but didn't have much success.

If we can get an updated version of `leinjacker` published, there seem to be [~50 public projects that use it](https://crossclj.info/ns/leinjacker/0.4.2/project.clj.html).

**Comment by MatthewDarling** [comment]
Turns out I had other plugins pulling in the older version of `leinjacker` instead of my new version. That's what caused the failures for `lein test leinjacker.deps-test` and the problems I had running `lein codox` with it. Both are working now with an empty `profiles.clj` :)

**Comment by MatthewDarling** [comment]
Like #166, this was fixed by #182. Leinjacker 0.4.3 included the necessary code changes and was released earlier this year.

(Yes, I know this issue is stale by two years, but so is the code I've been trying to rebuild today...)

**Comment by weavejester** [comment]
Thanks. If it's fixed, I'll close it.
